### PR TITLE
[@types/next] Add NextComponentType akin to React.ComponentType.

### PR DESCRIPTION
**Pull request by joaovieira**
This PR specifically adds `NextComponentType` and `AppComponentType` so they can be used in HOCs the same way we normally use `React.ComponentType`.

`NextContext.Component` is not a `React.ComponentType` but a `NextComponentType` instead, given it contains the Next.js specific lifecycle method `getInitialProps`:

screen shot 2018-09-19 at 16 31 28

The `NextComponentType` is also needed for HOCs that receive Next.js page or app components. See [tests](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/29021/files#diff-b201442b6dd5bf2932ebee4743ea75c4R54) for examples on how to add types for those. E.g. this is similar to what `withRedux` does:

```tsx
const withExample = <P extends {}>(Page: NextComponentType<P & WithExampleProps, P>) =>
    class extends React.Component<P & WithExampleHocProps> {
        test: string;

        static async getInitialProps(ctx: NextContext) {
            const pageProps = Page.getInitialProps && (await Page.getInitialProps(ctx));

            // tslint:disable-next-line prefer-object-spread
            return Object.assign({}, pageProps, { test: "test" });
        }

        constructor(props: P & WithExampleHocProps) {
            super(props);
            this.test = props.test;
        }

        render() {
            return <Page {...this.props} example={this.test} />;
        }
    };
```

`<P>` is inferred to match the wrapped component `Initial Props`. The HOC component then adds `WithExampleHocProps` to itself and finally renders the wrapped component with its own initial props plus `WithExampleProps`.

This also updates the template types for `NextStatelessComponent` and `App` to receive a `Context` type instead of the `Query` for consistency with all the other generics and given that `NextContext` can receive a `Query` param already. This allows to define your own context type as well as your own query schema. /cc @kaw2k 

Finally, renamed and deprecated some types for consistency. E.g. `NextContext`/`NextAppContext`/`NextDocumentContext` and `AppProps`/`DocumentProps`.

----

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: N/A, see test examples
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/next/app.d.ts [hunk 1]
```diff
@@ -1,19 +1,46 @@
 import * as React from "react";
-import { NextContext } from ".";
+import { NextContext, NextComponentType } from ".";
 import { RouterProps, DefaultQuery } from "./router";
 
-export interface AppComponentProps<Q = DefaultQuery> {
-    Component: React.ComponentType<any>;
+// Deprecated
+export type AppComponentProps = AppProps;
+export type AppComponentContext = NextAppContext;
```
Comment (joaovieira) [inline types/next/app.d.ts]: Renamed for consistency but still exported for backward compatibility.

File: types/next/index.d.ts [hunk 2]
```diff
@@ -44,19 +50,11 @@ declare namespace next {
...
-    }
-
-    type UrlLike = url.UrlObject | url.Url;
-
     /**
      * Next.js config schema.
      * https://github.com/zeit/next.js/blob/6.1.1/server/config.js#L10
      */
-    interface ServerConfig {
+    interface NextConfig {
```
Comment (joaovieira) [inline types/next/index.d.ts]: This is what Next.js calls this (https://github.com/zeit/next.js/blob/6.1.2/server/config.js#L10).

File: types/next/app.d.ts [hunk 3]
```diff
@@ -1,19 +1,46 @@
...
+// End Deprecated
+
+/**
+ * Context passed to App.getInitialProps.
+ * Component is dynamic - cannot infer type.
+ *
+ * @template Q Query object schema.
+ */
+export interface NextAppContext<Q = DefaultQuery> {
```
Comment (joaovieira) [inline types/next/app.d.ts]: Reordered interfaces for consistency aligned with the order in `Document`. The only difference is the `Component: NextComponentType<any, any, NextContext<Q>>` type in both.

**Comment by scottdj92** [comment]
👌

**Comment by joaovieira** [comment]
@resir014 I say a separate one. There's things that need to be done for v7 - e.g. remove the entire `next/dynamic` types. There's also some refactor to do on the `DocumentProps` and `RenderPageResponse` that can go in there.

**Review by resir014 — APPROVED** [review]

**Pull request merged**